net: drivers: wifi: SimpleLink WiFi Offload Driver (wifi_mgmt only)

GAnthony opened the pull request:
Initiate a SimpleLink WiFi Driver, implemented to the WiFi management
offload APIs for scan, connect, disconnect.

Also registers the DHCP-obtained IPv4 address upon connect.

This was validated on a cc3220sf_launchxl using the wifi
shell module from the Zephyr shell_module sample.   
A commit for that sample test is also included.

This is a replacement for the closed PR #6466.

Commit:
net: drivers: wifi: SimpleLink WiFi Offload Driver (wifi_mgmt only)

Initiate a SimpleLink WiFi Driver, implemented to the WiFi management
offload APIs for scan, connect, disconnect.

Also registers the DHCP-obtained IPv4 address upon connect.

This was validated on a cc3220sf_launchxl using the wifi
shell module from the Zephyr shell_module sample.

Signed-off-by: Gil Pitney <gil.pitney@linaro.org>

jukkar commented on drivers/wifi/simplelink.c:
```diff
@@ -0,0 +1,182 @@
+	case NET_EVENT_WIFI_CMD_CONNECT_RESULT:
+		/* Only get this event if connect succeeds: */
+		wifi_mgmt_raise_connect_result_event(simplelink_data.iface,
+							status);
```
Wrong indentation

jukkar commented on drivers/wifi/simplelink.c:
```diff
@@ -0,0 +1,182 @@
+			k_yield();
+			index++;
+		}
+		/* Sending a NULL entry indicates e/o results, and
```
Please put empty line after }

jukkar commented on drivers/wifi/simplelink_support.c:
```diff
@@ -0,0 +1,616 @@
+	ASSERT_ON_ERROR(retval, NETAPP_ERROR);
+
+	/* Disable ipv6 */
+	if_bitmap = !(SL_NETCFG_IF_IPV6_STA_LOCAL |
```
Is there some bigger issue with IPv6, or do we disable this because this is just not tested properly?
I mean would it make sense to enable IPv6 if user has enabled CONFIG_NET_IPV6 , now we always disable it?

jukkar commented on drivers/wifi/simplelink_support.c:
```diff
@@ -0,0 +1,616 @@
+		       BSSID_LEN_MAX);
+
+		SYS_LOG_INF("\n[WLAN EVENT] STA Connected to the AP: %s, "
+		       "BSSID: %x:%x:%x:%x:%x:%x",
```
Indentation issue

GAnthony replied:
got it.

jukkar commented on drivers/wifi/simplelink_support.c:
```diff
@@ -0,0 +1,616 @@
+		memcpy(sl_conn.bssid, wlan_event->Data.Connect.Bssid,
+		       BSSID_LEN_MAX);
+
+		SYS_LOG_INF("\n[WLAN EVENT] STA Connected to the AP: %s, "
```
No need to put \n to syslog prints, the output will look funny in practice if you have \n in the output.

GAnthony replied:
I generally agree with the principle of not putting newlines in log print out, especially if the underlying logger appends a '\n'.
But in this case, the SYS_LOG message prefix is so long  (eg: "dev/simplelink [INF] SimpleLinkWlanEventHandler: ") that the important information (the [WLAN EVENT] + SSID and IP addresses) end up wrapping around the console (typically set around 40 or so characters), so that it actually becomes very difficult to read.
Adding the '\n' at the start keeps the SSID names and IP addresses within the same line, and so much more readable.

jukkar commented on drivers/wifi/simplelink_support.c:
```diff
@@ -0,0 +1,616 @@
+		 */
+		if (SL_WLAN_DISCONNECT_USER_INITIATED ==
+		    event_data->ReasonCode) {
+			SYS_LOG_INF("\n[WLAN EVENT] "
```
Same thing about \n, just put everything in one line.

jukkar commented on drivers/wifi/simplelink_support.c:
```diff
@@ -0,0 +1,616 @@
+				    event_data->Bssid[5]);
+			sl_conn.error = 0;
+		} else {
+			SYS_LOG_ERR("\n[WLAN ERROR] "
```
Ditto

jukkar commented on drivers/wifi/simplelink_support.c:
```diff
@@ -0,0 +1,616 @@
+				    event_data->ReasonCode);
+			sl_conn.error = event_data->ReasonCode;
+		}
+		memset(&(sl_conn.ssid), 0x0, sizeof(sl_conn.ssid));
```
Empty line after }

GAnthony replied:
fixed.

jukkar commented on drivers/wifi/simplelink_support.c:
```diff
@@ -0,0 +1,616 @@
+	case SL_WLAN_EVENT_STA_ADDED:
+		memcpy(&(sl_conn.bssid), wlan_event->Data.STAAdded.Mac,
+		       SL_WLAN_BSSID_LENGTH);
+		SYS_LOG_INF("\n[WLAN EVENT] STA was added to AP: "
```
Remove \n

jukkar commented on drivers/wifi/simplelink_support.c:
```diff
@@ -0,0 +1,616 @@
+	case SL_WLAN_EVENT_STA_REMOVED:
+		memcpy(&(sl_conn.bssid), wlan_event->Data.STAAdded.Mac,
+		       SL_WLAN_BSSID_LENGTH);
+		SYS_LOG_INF("\n[WLAN EVENT] STA was removed from AP: "
```
Ditto

jukkar commented on drivers/wifi/simplelink_support.c:
```diff
@@ -0,0 +1,616 @@
+		/* Gateway IP address */
+		sl_conn.gateway_ip = event_data->Gateway;
+
+		SYS_LOG_INF("\n[NETAPP EVENT] IP set to: IPv4=%d.%d.%d.%d, "
```
Ditto

jukkar commented on drivers/wifi/simplelink_support.c:
```diff
@@ -0,0 +1,616 @@
+			  netapp_event->Data.IpAcquiredV6.Ip[i];
+		}
+
+		SYS_LOG_INF("[NETAPP EVENT] IP Acquired: IPv6=");
```
In practice splitting syslog prints will not work very well, the output will look very weird because each syslog call will print the header.  You could just call net_addr_ntop() to get the string presentation of IPv6 address and then print everything in one go.

So perhaps something like
```
#if SYS_LOG_LEVEL >= 3
{
    char ipv6_addr[NET_IPV6_ADDR_LEN];

    net_addr_ntop(AF_INET6, sl_conn.ipv6_addr, ipv6_addr, sizeof(ipv6_addr));
    SYS_LOG_INF(".....");
}
#endif
```

jukkar replied:
Noticed error in my comment, it should have been "#if SYS_LOG_LEVEL >= SYS_LOG_LEVEL_INFO" instead (not good idea to use hardcoded numbers here).

GAnthony replied:
Good suggestion, I missed that porting an older version of the handler from printf's.

jukkar commented on drivers/wifi/simplelink_support.c:
```diff
@@ -0,0 +1,616 @@
+		SET_STATUS_BIT(nwp.status, STATUS_BIT_IP_ACQUIRED);
+
+		sl_conn.sta_ip = netapp_event->Data.IpLeased.IpAddress;
+		SYS_LOG_INF("\n[NETAPP EVENT] IP Leased to Client: "
```
Remove \n

jukkar commented on drivers/wifi/simplelink_support.c:
```diff
@@ -0,0 +1,616 @@
+		break;
+
+	case SL_DEVICE_EVENT_DROPPED_NETAPP_IP_RELEASED:
+		SYS_LOG_INF("\n[NETAPP EVENT] IP is released.");
```
Ditto

jukkar commented on drivers/wifi/simplelink_support.c:
```diff
@@ -0,0 +1,616 @@
+		break;
+
+	default:
+		SYS_LOG_ERR("\n[NETAPP EVENT] Unexpected event [0x%lx]",
```
and here too

jukkar commented on drivers/wifi/simplelink_support.c:
```diff
@@ -0,0 +1,616 @@
+		return;
+	}
+
+	SYS_LOG_INF("\n[GENERAL EVENT] - ID=[%d] Sender=[%d]",
```
Ditto

jukkar commented on drivers/wifi/simplelink_support.c:
```diff
@@ -0,0 +1,616 @@
+
+	switch (fatal_err_event->Id) {
+	case SL_DEVICE_EVENT_FATAL_DEVICE_ABORT:
+		SYS_LOG_ERR("\n[ERROR] - FATAL ERROR: "
```
Can you check all sys_log lines and fix the \n, there are just too many of them to comment everyone.

GAnthony replied:
Same reason, it actually looks better with '\n' prefixes given the long SYS_LOG message prefix (due to the long CamelCase names of the "SimpleLink_XYZ_EventHander"s).

jukkar commented on drivers/wifi/simplelink_support.c:
```diff
@@ -0,0 +1,616 @@
+					    CONFIG_WIFI_SIMPLELINK_SCAN_COUNT,
+					    &nwp.net_entries[0]);
+	}
+	SYS_LOG_DBG("sl_WlanGetNetworkList: %d;  retries: %d", ret, trials);
```
Empty line after }

GAnthony replied:
fixed.

jukkar commented on ext/hal/ti/simplelink/Kconfig:
```diff
@@ -11,7 +11,7 @@ config SIMPLELINK_HOST_DRIVER
 	depends on HAS_CC3220SDK
 	depends on MULTITHREADING
-	depends on NEWLIB_LIBC
+	select NEWLIB_LIBC
```
Why do we need newlib here?

GAnthony replied:
Well, actually, the SimpleLink host driver fails to compile without newlib. 
Some missing symbols: uint_least8_t (from stdint.h), and then when sockets are being used, it needs fcntl.h.
It's the same issue with Zephyr sockets (all samples are compiling with NEWLIB_LIBC).
Given cc3220 has 1Mb flash and 256K SRAM, it's not a huge issue - but with other platforms, it might be.

pfalcon replied:
> Well, actually, the SimpleLink host driver fails to compile without newlib.
Some missing symbols: ...

@GAnthony: This becomes a recurring problem (e.g. just had it with looking into cypress psoc6 work started). We need to submit tickets for these and fix whenever possible.

jukkar reviewed (changes requested):
Looks good, minor nitpicks. I will also let @tbursztyka to review these.

GAnthony reviewed (commented):
Thanks for the feedback.

GAnthony commented:
@erwango, invite to take a look, if also writing a wifi driver soon.

tbursztyka commented on drivers/wifi/simplelink.c:
```diff
@@ -0,0 +1,183 @@
+		/* Only get this event if connect succeeds: */
+		wifi_mgmt_raise_connect_result_event(simplelink_data.iface,
+						     status);
+		k_yield(); /* Needed to fire net_event notifications: */
```
That's actually not true.

The event will be fired, with or without the k_yield(). You are confusing with the scan result event case where using k_yield() helps to avoid loosing events if the system is over-loaded, knowing that there might be numerous scan result events, that's all. Here it's a lone event, unlikely to be lost.

Remove this

GAnthony replied:
> Here it's a lone event, unlikely to be lost.

Actually, during testing, when providing an SSID with an invalid security key, I found that the offload chip continues to retry the connection.   It has an "auto retry" feature.   So, in this case, disconnect events are generated in a higher priority thread than the net mgmt thread, and so the wifi shell never sees any disconnect events.

There may be other similar cases with these offload chips which may not have been originally expected.

But in general, since the net event mgmt system is lossy, there is no guarantee that events will get delivered.  So until net mgmt framework is re-architected to guarantee events are not dropped, I would argue that doing a k_yield() after  each wifi_mgmt_raise_[x]_event() call is mandatory.

So, to be more exact, the comment should say something like "the k_yield() is needed to *ensure* the notification is not dropped."

tbursztyka replied:
> But in general, since the net event mgmt system is lossy, there is no guarantee that events will get delivered. So until net mgmt framework is re-architected to guarantee events are not dropped, I would argue that doing a k_yield() after each wifi_mgmt_raise_[x]_event() call is mandatory.

No way. net mgmt is meant for user targeted events, so there is just _no point_ in this. This is just a technical non sense.

If you have a solution, please go fix Netlink in Linux as well.

GAnthony replied:
Removed.

tbursztyka commented on drivers/wifi/simplelink.c:
```diff
@@ -0,0 +1,183 @@
+		/* Could be during a connect, disconnect, or async error: */
+		wifi_mgmt_raise_disconnect_result_event(simplelink_data.iface,
+							status);
+		k_yield(); /* Needed to fire net_event notifications: */
```
same

tbursztyka commented on drivers/wifi/simplelink.c:
```diff
@@ -0,0 +1,183 @@
+		 */
+		cb(iface, status, NULL);
+	}
+	return status;
```
This is not implemented according to API. 

Scan raises a scan, but should return immediately returning 0 on success (scan has been raised) or an error. 

You'll have to get own driver thread to actually call the callback.

GAnthony replied:
This scan API is forcing each offload driver, which gets scan results synchronously, to create a thread (or other complex mechanism) to fit into an asynchronous callback based scan API.

I still recommend the API be designed for ease of use, and to fit the most common use case (i.e., the first WiFi devices are these offload chips).

tbursztyka replied:
We already had this conversation about WiFi scan which is really an asynchronous mechanism. And also: we _really_ don't care about what are offloaded chip doing. Up to offloaded driver to fit with requested behaviour from wifi_mgmt.

GAnthony replied:
@tbursztyka, OK, I used a system workqueue to make scan behave more asynchronously.
It was a good opportunity to learn about Zephyr workqueues, and it actually worked out quite nicely.

tbursztyka commented on drivers/wifi/simplelink.c:
```diff
@@ -0,0 +1,183 @@
+	status = (num_results_or_err < 0 ? -EIO : 0);
+
+	/* Now iterate over the table, and call the scan_result callback. */
+	if (cb) {
```
callback is always provided. no need to test it validity

GAnthony replied:
There is no contract in `wifi_mgmt.h:net_wifi_mgmt_offload{... scan() ...}` to specify that cb is always valid.
The current code in `wifi_mgmt.c:wifi_scan()` ensures cb is valid, until someone changes that code later on, causing wifi drivers' scan functions to crash.
So, I can change the check to an assert() to test this hidden assumption.

tbursztyka replied:
yeah well there won't be numerous implementation of wifi_mgmt: only one. So what it does _is_ the contract (all of our APIs lacks of exhaustive documentation after all)

tbursztyka commented on drivers/wifi/simplelink.c:
```diff
@@ -0,0 +1,183 @@
+static struct simplelink_data simplelink_data;
+
+/* Handle connection events from the SimpleLink Event Handlers: */
+static void simplelink_wifi_cb(u32_t mgmt_event, struct sl_connect_state *conn)
```
Use your own code, not mgmt_event, to tell your callback what it should do here.

You driver is not listening to NET_EVENT_WIFI_CMD_(DIS)CONNECT_RESULT, nor NET_EVENT_IPV4_ADDR_ADD (not to say about the fact you are mixing up event cmd enums and macros here.)

GAnthony replied:
OK, good point.

tbursztyka replied:
rename mgmt_event to event.

GAnthony replied:
OK, done.

GAnthony commented on drivers/wifi/simplelink.c:
```diff
@@ -0,0 +1,183 @@
+static struct simplelink_data simplelink_data;
+
+/* Handle connection events from the SimpleLink Event Handlers: */
+static void simplelink_wifi_cb(u32_t mgmt_event, struct sl_connect_state *conn)
```
OK, good point.

GAnthony commented:
@tbursztyka, thanks for taking the time to review.

GAnthony commented:
recheck

tbursztyka commented on drivers/wifi/simplelink_support.c:
```diff
@@ -0,0 +1,599 @@
+	s32_t mode = -1;
+	u32_t if_bitmap = 0;
+	SlWlanScanParamCommand_t scan_default = { 0 };
+	SlWlanRxFilterOperationCommandBuff_t rx_filterid_mask = { { 0 } };
```
tiny point, we usually do the invert here: first declaring variable pre-assigned, and then only the lone declaration.

GAnthony replied:
OK

tbursztyka commented on drivers/wifi/simplelink_support.c:
```diff
@@ -0,0 +1,599 @@
+{
+	u32_t i = 0;
+
+	SlIpV4AcquiredAsync_t *event_data = NULL;
```
revert the order of these 2 declarations. And don't pre-assign 'i': your for loop is doing it anyway

GAnthony replied:
Thanks.

tbursztyka commented on drivers/wifi/simplelink_support.c:
```diff
@@ -0,0 +1,599 @@
+int _simplelink_connect(struct wifi_connect_req_params *params)
+{
+	SlWlanSecParams_t secParams = { 0 };
+	long lretval = 0;
```
seems like a useless assignment to me

tbursztyka commented on drivers/wifi/simplelink_support.c:
```diff
@@ -0,0 +1,599 @@
+
+int _simplelink_disconnect(void)
+{
+	long lretval = 0;
```
ditto

GAnthony replied:
Got it.

Commit:
samples: net: wifi: Add a cc3220sf_launchxl conf file

This enables testing of the scan, connect, and disconnect
wifi_mgmt functions of the wifi shell module for
the cc3220sf_launchxl.

Signed-off-by: Gil Pitney <gil.pitney@linaro.org>

GAnthony commented:
recheck

GAnthony commented:
@tbursztyka, @jukkar, are there any other requested changes required to allow this PR to be merged?

tbursztyka commented:
well, if it's only wifi net mgmt, it still requires the net_offload struct to be populated and working so user can actually get a network connection and use it. It would be ok to get that PR pushed, but it would be a bit useless from usage point of view.

GAnthony commented:
> but it would be a bit useless from usage point of view

Actually, it would be very useful.  The plan would be to first update PR #5028 to work with this driver, then add some default provisioning.  
We have a customer already using WiFi in Zephyr on this chip, but bypassing the Zephyr IP stack.  At least with this PR, we can have some start of an integration of the WiFi management functionality, then socket offload can come next (in whatever form that will eventually take).

nashif commented:
+1 for pushing this and getting everything else related to this in the 1.13 timeframe, this has been in the queue for way too long.

jukkar reviewed (approved):
LGTM

GAnthony commented:
Many Thanks!
